examples/src/main/scala/io/chrisdavenport/shellfish/contacts/core/ContactManager.scala

```diff
@@ -0,0 +1,142 @@
+
+trait ContactManager {
+
+  def addContact(contact: Contact): IO[Username]
+
+  def removeContact(username: Username): IO[Unit]
+
+  def searchId(username: Username): IO[Option[Contact]]
+  def searchName(name: Name): IO[List[Contact]]
+  def searchEmail(email: Email): IO[List[Contact]]
+  def searchNumber(number: PhoneNumber): IO[List[Contact]]
+
+  def getAll: IO[List[Contact]]
+
+  def updateContact(username: String)(modify: Contact => Contact): IO[Contact]
+
+}
+
+object ContactManager {
+  def impl(bookPath: Path): ContactManager = new ContactManager {
+
+    private def parseContact(contact: String): Either[Exception, Contact] =
+      contact.split('|') match {
+        case Array(id, firstName, lastName, phoneNumber, email) =>
+          Contact(id, firstName, lastName, phoneNumber, email).asRight
+
+        case _ => new Exception(s"Invalid contact format: $contact").asLeft
+      }
+
+    private def encodeContact(contact: Contact): String =
+      s"${contact.username}|${contact.firstName}|${contact.lastName}|${contact.phoneNumber}|${contact.email}"
+
+    override def addContact(contact: Contact): IO[Username] = {
+      bookPath.readLines.flatMap { lines =>
+        lines.find(_.startsWith(contact.username)) match {
+          case Some(_) =>
+            ContactFound(contact.username).raiseError[IO, Username]
+
+          case None =>
+            lines.isEmpty
+              .pure[IO]
+              .ifM(
+                bookPath.write(encodeContact(contact)).as(contact.username),
+                bookPath.appendLine(encodeContact(contact)).as(contact.username)
+              )
+        }
+      }
+    }
+
+    override def removeContact(username: Username): IO[Unit] =
+      for {
+        contacts <- bookPath.readLines
+        _ <- bookPath.writeLines(
+          contacts.filterNot(_.startsWith(username))
+        )
+      } yield ()
+
+    override def searchId(username: Username): IO[Option[Contact]] =
+      for {
+        contacts <- bookPath.readLines
+      } yield contacts
+        .find(_.startsWith(username))
+        .flatMap(parseContact(_).toOption)
+
+    override def searchName(name: Name): IO[List[Contact]] =
+      for {
+        contacts <- bookPath.readLines.map(_.filter(_.contains(name)))
+        parsed <- contacts.traverse(contact =>
+          IO.fromEither(parseContact(contact))
+        )
+      } yield parsed
+
+    override def searchEmail(email: Email): IO[List[Contact]] =
+      for {
+        contacts <- bookPath.readLines.map(_.filter(_.contains(email)))
+        parsed <- contacts.traverse(contact =>
+          IO.fromEither(parseContact(contact))
+        )
+      } yield parsed
+
+    override def searchNumber(number: PhoneNumber): IO[List[Contact]] =
+      for {
+        contacts <- bookPath.readLines.map(_.filter(_.contains(number)))
+        parsed <- contacts.traverse(contact =>
+          IO.fromEither(parseContact(contact))
+        )
+      } yield parsed
```

Comment:
Using [this](https://github.com/davenverse/shellfish/pull/86/files#r1730017009) version of `getAll`. It's less efficient, but it's simpler, and that's what we're aiming for.

```suggestion
    override def searchId(username: Username): IO[Option[Contact]] =
      getAll.map(contacts => contacts.find(_.username === username))

    override def searchName(name: Name): IO[List[Contact]] =
      getAll.map(contacts =>
        contacts.filter(c => c.firstName === name || c.lastName === name)
      )

    override def searchEmail(email: Email): IO[List[Contact]] =
      getAll.map(contacts => contacts.filter(_.email === email))

    override def searchNumber(number: PhoneNumber): IO[List[Contact]] =
      getAll.map(contacts => contacts.filter(_.phoneNumber === number))
```